### DISCUSSION TITLE
I2VGenXLPipeline - missing components?

### DISCUSSION THREAD

**Discussion by dhaivat1729** [huggingface/diffusers#7952] [Q&A]
Hi everyone,

I was playing with [I2VGenXLPipeline](https://arxiv.org/pdf/2311.04145). Here is corresponding [Huggingface implementation.](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py). I saw some discrepancy between method described in the paper and this implementation. Can someone help me in checking if my understanding is correct. 

In the paper, they have the following diagram: 
![Screenshot 2024-05-15 at 2 00 28 PM](https://github.com/huggingface/diffusers/assets/6570814/fe96ba73-4af3-4169-b33e-5ffa8e95f99d)

- According to this diagram, in the base stage, they have `D.Enc.` and `G.Inc`, however, I only see CLIP in the implementation [here](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py#L120).  

- Similarly, in implementation, I observe that text embeddings are passed to the LDM of base stage ([this line](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py#L684)), however, as per the diagram, text is only passed in refinement stage. 

- In refinement stage, there is LDM, however, in implementation, I see that low dimensional [video latent is passed to VAE decoder](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py#L718) to generate high dimensional video, I do not see any reverse diffusion process. 

Can anyone tell me if my understanding is correct for this code? I wanted to access intermediate low dimensional video, which comes at the end of base stage, but I don't know how to exactly access it. Can anyone tell me how to access that representation? I would appreciate it.

**Comment by yiyixuxu** [comment]
cc @sayakpaul

**Comment by sayakpaul** [comment]
Thanks for bringing this to our attention. We followed the original [implementation code](https://github.com/damo-vilab/i2vgen-xl.git) and ensured the same outputs could be obtained from the two implementations.   

1. They're very likely used only during training and NOT during inference (similar to other models like https://arxiv.org/abs/2306.00637). You can verify this from the [original implementation code](https://github.com/damo-vilab/i2vgen-xl.git) for inference as well. 
2. Again, it could very well be just for training and not for inference. 

> In refinement stage, there is LDM, however, in implementation, I see that low dimensional [video latent is passed to VAE decoder](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py#L718) to generate high dimensional video, I do not see any reverse diffusion process.

I think there's a misunderstanding here. LDM is the entire process:

* You start with a low-dimensional noise vector.
* You then iteratively refine the noise vector with a denoiser (a UNet in this case) that is conditioned on text, timestep of the noise scheduler, etc. 
* You pass the refined latent (aka the denoised noise vector) to the decoder of a VAE to obtain the final output. 

So, to answer your question, all the above steps are being done in https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/i2vgen_xl/pipeline_i2vgen_xl.py. 

Also, you can access the intermediate latents at the end of the iterative process by setting `output_type="latents"` in the pipeline call. 

Hope that makes sense.

**Reply by dhaivat1729** [reply]
Hi @sayakpaul 

Thank you for your response. 

I will check their original implementation. According to their figure, there appears to be 2 LDM systems, one in base stage and one is refinement stage. 
However, prompt is only passed to refinement stage as per the figure. However, right now we are passing prompt along with noise vector, which was the source of confusion. 

In current implementation, denoising process is happening only once, and output of denoising is low dimensional video (which is end of the base stage), and then prompt is passed along with this low dimensional video in the refinement stage. However, currently, we have output of denoising a low dimensional video, which has already seen prompt. 

I will contact the authors of the paper to see if there is any discrepancy.

**Comment by Steven-SWZhang** [comment]
Hello,
Thank you for your interest in our work on I2VGen-XL. In fact, I2VGen-XL comes in several different versions. We have not released the two-stage model here because the two-stage model does not sufficiently retain the content of the input image (you can observe this carefully in this fig); the two-stage model is relatively more complex with a larger number of parameters, making it less suitable for academic use. Therefore, we have released our single-stage model instead, which can fully preserve the content of input images and has a simpler overall structure. If you are looking to access our second-stage model, we have actually already made it open-source, please refer [here](https://github.com/ali-vilab/VGen/blob/main/tools/inferences/inference_sr600_entrance.py). If you are just interested in obtaining the intermediate low-resolution video, you may need try the implementation available [here on modelscope](https://github.com/modelscope/modelscope/blob/master/modelscope/models/multi_modal/image_to_video/image_to_video_model.py).

**Reply by dhaivat1729** [reply]
Thank you @Steven-SWZhang , that makes sense. I will try to run modelscope version to get low res video, thank you.

**Reply by dhaivat1729** [reply]
@sayakpaul 
I don't know if it makes sense or not, but in documentation, we can explicitly state that the HF implementation is a single stage variant of I2VGen-XL, as the main figure in the paper is 2 stage, it could be made clearer. Just a suggestion :)

**Reply by sayakpaul** [reply]
Sure, feel free to drop a PR and will get that sorted.

**Reply by dhaivat1729** [reply]
@sayakpaul sure, I would love to make a PR, can you tell which all files should I look at to make the changes? I am new to the repository and I am not fully sure. I am sure the PR itself will not take time but I want to do it right way. 

One place I can think of is docstring of I2vGen class itself, anywhere else?

**Reply by sayakpaul** [reply]
Thank you!

I think you could perform the edit in https://huggingface.co/docs/diffusers/en/api/pipelines/i2vgenxl, the source is here: https://github.com/huggingface/diffusers/blob/main/docs/source/en/api/pipelines/i2vgenxl.md. 

Does that help?

**Reply by dhaivat1729** [reply]
@sayakpaul yes, it does. Thank you for your response, I have submitted PR for the same, this is my first PR to any HF repo, so please provide your suggestions and I will make amendments. I just added 1 more point in Notes section of the docs. 
PR link: [https://github.com/huggingface/diffusers/pull/8282](https://github.com/huggingface/diffusers/pull/8282)
